### DISCUSSION TITLE
Using a numeric value as a key for a database connection in config file will result in it not being loaded with that name.

### DISCUSSION THREAD

**Discussion by ItsClassified** [laravel/framework#52956] [Ideas]
### Laravel Version

11.23.5

### PHP Version

8.3.11

### Database Driver & Version

MySQL

### Description

When using numeric values in the database.php config file Laravel will disregard the names and load them starting from 0, and increasing for each connection after that with a numeric value as name.

@ollieread found that commenting out row 127 in the following file makes it load in again: https://github.com/laravel/framework/blob/11.x/src/Illuminate/Foundation/Bootstrap/LoadConfiguration.php#L103

The issue appears to be on L103 using `array_merge`

### Steps To Reproduce

Change a connection name to a numeric string, for example: '12345' and try to access it using DB::connection.

**Comment by ollieread** [comment]
Just to add a bit more, the issue is when it tries to merge `database.connections`, as PHPs `array_merge()` is doing some annoying type juggling and assuming keys that are numeric strings, are, in fact, just numeric, and it's resetting those keys.

**Comment by crynobone** [comment]
I don't think we have considered using an integer key as a valid configuration key. Moving this into Ideas for now and feel free to discuss this further.

**Reply by ItsClassified** [reply]
Thanks for taking the time @crynobone , the reason why I made the issue is because this worked in every version before Laravel 11, and broke our application after updating to Laravel 11. Since it broke after updating its fair to say its an issue, given that it is not disclosed in the upgrade section of Laravel 11 that it no longer supports using a numeric value as a key for a config file...

**Reply by crynobone** [reply]
> Since it broke after updating its fair to say its an issue, given that it is not disclosed in the upgrade section of Laravel 11 that it no longer supports using a numeric value as a key for a config file...

It was never considered as an expected usage, so you cannot say that it should be disclosed. It would be different if our official documentation or skeleton uses integer as key before and starting with Laravel 11 it no longer supported then yes your statement is valid.

**Reply by ItsClassified** [reply]
Okay, however I feel like when u use '564321' => as a key, (and not just 564321 =>  as an int), it should be treated as an string and work as expected. Im obvioulsy biased because our entire application is based on having keys like '564321' as database connections, and upgrading to Laravel 11 is gonna be a pain in the ass otherwise...

**Reply by donnysim** [reply]
One more reason why mergeable configs are horrible and I disabled it instantly.

**Reply by ollieread** [reply]
> > Since it broke after updating its fair to say its an issue, given that it is not disclosed in the upgrade section of Laravel 11 that it no longer supports using a numeric value as a key for a config file...
> 
> 
> 
> It was never considered as an expected usage, so you cannot say that it should be disclosed. It would be different if our official documentation or skeleton uses integer as key before and starting with Laravel 11 it no longer supported then yes your statement is valid.

Hey @crynobone, I think there may have been a slight misunderstanding. This issue is not that you can't use integer keys, it's that you can't use string keys that PHP may interpret as a number. So we aren't facing a bug where you can't use ints, but rather you can't use some strings.

**Reply by crynobone** [reply]
![Screenshot_20240927_161316_Chrome](https://github.com/user-attachments/assets/973ed04d-395d-40e8-bd7d-46dc71f41cec)

This is from PHP official documentation.

**Reply by ollieread** [reply]
> ![Screenshot_20240927_161316_Chrome](https://private-user-images.githubusercontent.com/172966/371450716-973ed04d-395d-40e8-bd7d-46dc71f41cec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0Mjc2NTAsIm5iZiI6MTcyNzQyNzM1MCwicGF0aCI6Ii8xNzI5NjYvMzcxNDUwNzE2LTk3M2VkMDRkLTM5NWQtNDBlOC1iZDdkLTQ2ZGM3MWY0MWNlYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkyN1QwODU1NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNGU4YmZlODE4NmI5YTkzMTI1MDE2NDMxMzliNTJmMWRmMjEzNzNiYWQwMjg1NDVlNjhmYWVjY2FhMmMxOGQ4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.zqWNXwIi5OQsbNsSePxUdk5mVaPuWJlcWa2IyEpqjqM)
> 
> This is from PHP official documentation.

Can you provide a link to this, as I'm not sure if it's relevant? PHP lets you use numeric strings as array keys, and you can access them as both numbers and strings. The problem is that `array_merge()` detects that they're numeric strings, and just resets them to be incremental.

**Reply by crynobone** [reply]
https://www.php.net/manual/en/language.types.array.php#language.types.array.syntax

**Reply by crynobone** [reply]
https://www.php.net/manual/en/language.types.array.php#language.types.array.syntax

**Reply by rodrigopedra** [reply]
> PHP lets you use numeric strings as array keys, and you can access them as both numbers and strings. 

Numeric strings, that represent valid integers, are cast to integers on both offset set and offset get operations.

That is different from using "numeric strings" as keys.

> This issue is not that you can't use integer keys, it's that **you can't use string keys that PHP may interpret as a number**.

That is the expected behavior, from PHP's perspective.

```php
<?php

$arr = ['0' => 'foo', '1' => 'bar'];

// - keys will be dumped as integers
// - \array_list($arr) will return true
\var_dump($arr, \array_is_list($arr));

$arr['03'] = 'baz';

// - only the new key will be quoted as a string
// - \array_list($arr) will return false
\var_dump($arr, \array_is_list($arr));
```

```text
array(2) {
  [0]=>
  string(3) "foo"
  [1]=>
  string(3) "bar"
}
bool(true)
array(3) {
  [0]=>
  string(3) "foo"
  [1]=>
  string(3) "bar"
  ["03"]=>
  string(3) "baz"
}
bool(false)
```

Reference: https://3v4l.org/8cei8

> So we aren't facing a bug where you can't use ints, but rather you can't use some strings.

If you believe this is a bug, I'd consider it'd be better to open a bug report to PHP.

**Reply by ollieread** [reply]
> If you believe this is a bug, I'd consider it'd be better to open a bug report to PHP.

@rodrigopedra When adding an entry to an array with a string key that is a valid integer, it is cast. When accessing that entry in the array, the key you provide is also cast, so you can access as either an `int` or a `string`. 

There is nothing wrong with this, at all. After all, this is how the language works.

The problem, is that Laravel merges several configuration values with their defaults, which overwrites these keys, ostensibly changing the name of the connection. I'm still yet to find the reasoning behind this merge, or its purpose, but either way, it changes the way you can define connections.

**Reply by rodrigopedra** [reply]
@ollieread I see now what the issue is.

Well, as @donnysim said, you can disable configuration merge, and use the `config:publish` command to publish any missing config file.

I wasn't a big fan of this change - regarding auto-merging configuration - on Laravel 11. But I guess this ship has sailed.

Another option, is to propose a PR to change how `Illuminate\Foundation\Bootstrap\LoadConfiguration@loadConfigurationFile` works, to use some sort of deep-merge algorithm instead of using `array_merge()`.
